File: sdks/python/src/attest/plugins.py
"""Attest plugin system — Protocol, Registry, entrypoint discovery, execution."""

from __future__ import annotations

import asyncio
import importlib.metadata
import logging
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Protocol

from attest._proto.types import Trace

if TYPE_CHECKING:
    from attest.client import AttestClient

logger = logging.getLogger("attest.plugins")


class AttestPlugin(Protocol):
    """Protocol that all Attest plugins must implement."""

    name: str
    plugin_type: str  # "adapter", "assertion", "judge", "reporter"

    def execute(self, trace: Trace, spec: dict[str, Any]) -> PluginResult:
        """Execute the plugin against a trace and return a result."""
        ...


@dataclass
class PluginResult:
    """Result returned by a plugin execution."""

    status: str
    score: float
    explanation: str
    metadata: dict[str, Any] | None = None


class PluginRegistry:
    """Stores registered plugins keyed by (plugin_type, name)."""

    def __init__(self) -> None:
        self._plugins: dict[str, dict[str, AttestPlugin]] = {}

    def register(self, name: str, plugin: AttestPlugin) -> None:
        """Register a plugin instance under its plugin_type and name."""
        plugin_type = plugin.plugin_type
        if plugin_type not in self._plugins:
            self._plugins[plugin_type] = {}
        self._plugins[plugin_type][name] = plugin

    def get(self, plugin_type: str, name: str) -> AttestPlugin | None:
        """Retrieve a registered plugin by type and name. Returns None if not found."""
        return self._plugins.get(plugin_type, {}).get(name)

    def list_plugins(self, plugin_type: str | None = None) -> list[str]:
        """List registered plugin names.

        If plugin_type is provided, list only plugins of that type.
        Otherwise list all plugins as "type/name" strings.
        """
        if plugin_type is not None:
            return list(self._plugins.get(plugin_type, {}).keys())
        result: list[str] = []
        for ptype, plugins in self._plugins.items():
            for pname in plugins:
                result.append(f"{ptype}/{pname}")
        return result


def load_entrypoint_plugins(registry: PluginRegistry) -> int:
    """Discover and load plugins from the 'attest.plugins' entry point group.

    Returns the number of plugins successfully loaded.
    """
    eps = importlib.metadata.entry_points(group="attest.plugins")
    loaded = 0
    for ep in eps:
        try:
            plugin: AttestPlugin = ep.load()
            registry.register(ep.name, plugin)
            loaded += 1
            logger.debug("Loaded plugin %r from entry point", ep.name)
        except Exception:
            logger.exception("Failed to load plugin %r from entry point", ep.name)
    return loaded


def register_plugin(registry: PluginRegistry, name: str, plugin: AttestPlugin) -> None:
    """Explicitly register a plugin instance into the registry."""
    registry.register(name, plugin)


async def execute_plugin_assertion(
    plugin: AttestPlugin,
    trace: Trace,
    spec: dict[str, Any],
    client: AttestClient,
    trace_id: str,
    assertion_id: str,
    timeout: float = 30.0,
) -> PluginResult:
    """Execute a plugin with a timeout and submit the result via client.

    Runs the synchronous plugin.execute() in a thread executor to avoid
    blocking the event loop. Submits the result to the engine via
    client.submit_plugin_result() after execution.

    Raises asyncio.TimeoutError if execution exceeds timeout seconds.
    """
    loop = asyncio.get_running_loop()

    result: PluginResult = await asyncio.wait_for(
        loop.run_in_executor(None, plugin.execute, trace, spec),
        timeout=timeout,
    )

    await client.submit_plugin_result(
        trace_id=trace_id,
        plugin_name=plugin.name,
        assertion_id=assertion_id,
        status=result.status,
        score=result.score,
        explanation=result.explanation,
    )

    return result

File: sdks/typescript/packages/core/src/plugins.ts
import type { Trace } from "./proto/types.js";
import type { AttestClient } from "./client.js";

export interface PluginResult {
  readonly status: string;
  readonly score: number;
  readonly explanation: string;
  readonly metadata?: Record<string, unknown>;
}

export interface AttestPlugin {
  readonly name: string;
  readonly pluginType: string; // "adapter" | "assertion" | "judge" | "reporter"
  execute(trace: Trace, spec: Record<string, unknown>): PluginResult | Promise<PluginResult>;
}

export class PluginRegistry {
  private readonly plugins = new Map<string, Map<string, AttestPlugin>>();

  register(name: string, plugin: AttestPlugin): void {
    const pluginType = plugin.pluginType;
    let typeMap = this.plugins.get(pluginType);
    if (typeMap === undefined) {
      typeMap = new Map();
      this.plugins.set(pluginType, typeMap);
    }
    typeMap.set(name, plugin);
  }

  get(pluginType: string, name: string): AttestPlugin | undefined {
    return this.plugins.get(pluginType)?.get(name);
  }

  listPlugins(pluginType?: string): string[] {
    if (pluginType !== undefined) {
      const typeMap = this.plugins.get(pluginType);
      return typeMap !== undefined ? [...typeMap.keys()] : [];
    }
    const result: string[] = [];
    for (const [ptype, typeMap] of this.plugins) {
      for (const pname of typeMap.keys()) {
        result.push(`${ptype}/${pname}`);
      }
    }
    return result;
  }
}

export async function executePluginAssertion(
  plugin: AttestPlugin,
  trace: Trace,
  spec: Record<string, unknown>,
  client: AttestClient,
  traceId: string,
  assertionId: string,
  timeoutMs = 30_000,
): Promise<PluginResult> {
  const resultPromise = Promise.resolve(plugin.execute(trace, spec));

  const timeoutPromise = new Promise<never>((_, reject) => {
    setTimeout(() => reject(new Error(`Plugin '${plugin.name}' timed out after ${timeoutMs}ms`)), timeoutMs);
  });

  const result = await Promise.race([resultPromise, timeoutPromise]);

  await client.submitPluginResult(
    traceId,
    plugin.name,
    assertionId,
    result.status,
    result.score,
    result.explanation,
  );

  return result;
}
